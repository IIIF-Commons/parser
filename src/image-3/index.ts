export * from './types';
export * from './parser/parse-image-server-from-id';
export * from './parser/parse-image-service-request';
export * from './parser/parse-image-service-url';
export * from './parser/parse-region-parameter';
export * from './parser/parse-rotation-parameter';
export * from './parser/parse-size-parameter';
export * from './profiles/profiles';
export * from './profiles/combine-profiles';
export * from './profiles/level-to-profile';
export * from './profiles/is-level-0';
export * from './profiles/supports';
export * from './profiles/supports-custom-sizes';
export * from './profiles/image-service-supports-format';
export * from './profiles/image-service-supports-request';
export * from './serialize/image-service-request-to-string';
export * from './serialize/image-service-request-info';
export * from './serialize/region-parameter-to-string';
export * from './serialize/rotation-parameter-to-string';
export * from './serialize/size-parameter-to-string';
export * from './utilities/canonical-service-url';
export * from './utilities/create-image-service-request';
export * from './utilities/extract-fixed-size-scales';
export * from './utilities/fixed-sizes-from-scales';
export * from './utilities/get-id';
export * from './utilities/get-type';
export * from './utilities/is-image-service';