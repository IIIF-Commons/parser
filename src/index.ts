// Exports the latest
export * from './presentation-3';
