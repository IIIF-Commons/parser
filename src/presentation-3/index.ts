export * from "./empty-types";
export * from "./normalize";
export * from "./traverse";
export * from "./serialize";
export * from "./serialize-presentation-2";
export * from "./serialize-presentation-3";
