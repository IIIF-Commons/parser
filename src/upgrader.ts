import { convertPresentation2 } from './presentation-2/upgrader';

export const upgrade = convertPresentation2;
