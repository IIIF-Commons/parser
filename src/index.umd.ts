import * as Presentation2 from "./presentation-2";
import * as Presentation3 from "./presentation-3";

export default {
  Presentation2,
  Presentation3,
};
