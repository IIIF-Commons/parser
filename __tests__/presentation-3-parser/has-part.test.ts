import { describe, test, expect } from 'vitest';
import hasPartManifest from '../../fixtures/presentation-3/has-part.json';
import { normalize, serialize, mergeEntities, serializeConfigPresentation3 } from '../../src';

describe('Has part issues', function () {
  test('merging 2 entities', () => {
    const first = {
      id: 'https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg',
      type: 'Image',
      format: 'image/jpeg',
    };
    const second = {
      id: 'https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg',
      type: 'Image',
      format: 'image/jpeg',
      height: 3024,
      width: 4032,
      service: [
        {
          id: 'https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen',
          profile: 'level1',
          type: 'ImageService3',
        },
      ],
    };

    const initial = mergeEntities(
      {
        id: 'https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg',
        type: 'Image',
      },
      first,
      { parent: { id: 'https://example.org/canvas-2' } }
    );
    expect(initial).toMatchSnapshot();
    const merged = mergeEntities(initial as any, second, { parent: { id: 'https://example.org/canvas-1' } });

    expect(merged).toMatchSnapshot();
  });

  test('Merging 2 services', () => {
    const item1 = {
      service: [
        {
          '@id': 'https://iiif.wellcomecollection.org/image/b1465782x_0001.jp2',
          '@type': 'ImageService2',
          profile: 'http://iiif.io/api/image/2/level1.json',
          width: 6036,
          height: 7161,
          id: 'https://iiif.wellcomecollection.org/image/b1465782x_0001.jp2',
        },
      ],
    };

    const item2 = {
      service: [
        {
          '@id': 'https://iiif.wellcomecollection.org/image/b1465782x_0001.jp2',
          '@type': 'ImageService2',
          profile: 'http://iiif.io/api/image/2/level1.json',
          width: 6036,
          height: 7161,
        },
      ],
    };

    const merged = mergeEntities(item1 as any, item2, { parent: { id: 'https://example.org/canvas-1' } });

    expect(merged).toMatchSnapshot();

  });

  test('Merging the same 2 entities', () => {
    const first = {
      id: 'https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg',
      type: 'Image',
      format: 'image/jpeg',
    };
    const second = {
      id: 'https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg',
      type: 'Image',
      format: 'image/jpeg',
    };

    const initial = mergeEntities(
      {
        id: 'https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg',
        type: 'Image',
      },
      first,
      { parent: { id: 'https://example.org/canvas-2' } }
    );
    expect(initial).toMatchSnapshot();
    const merged = mergeEntities(initial as any, second, { parent: { id: 'https://example.org/canvas-1' } });

    expect(merged).toMatchSnapshot();
  });

  test('Merging the same 2 entities and then different one', () => {
    const first = {
      id: 'https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg',
      type: 'Image',
      format: 'image/jpeg',
    };
    const second = {
      id: 'https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg',
      type: 'Image',
      format: 'image/jpeg',
    };
    const third = {
      id: 'https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg',
      type: 'Image',
      format: 'image/jpeg',
      height: 3024,
      width: 4032,
      service: [
        {
          id: 'https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen',
          profile: 'level1',
          type: 'ImageService3',
        },
      ],
    };

    const initial = mergeEntities(
      {
        id: 'https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg',
        type: 'Image',
      },
      first,
      { parent: { id: 'https://example.org/canvas-2' } }
    );
    expect(initial).toMatchSnapshot();
    const merged = mergeEntities(initial as any, second, { parent: { id: 'https://example.org/canvas-1' } });
    const merged2 = mergeEntities(merged as any, third, { parent: { id: 'https://example.org/canvas-3' } });

    expect(merged2).toMatchSnapshot();
  });

  test('Example manifest with thumbnail ID the same as the main resource', () => {
    const original = JSON.parse(JSON.stringify(hasPartManifest));
    const result = normalize(hasPartManifest);
    expect(result).toMatchSnapshot();

    const reserialized = serialize(
      {
        mapping: result.mapping,
        entities: result.entities,
        requests: {},
      },
      result.resource,
      serializeConfigPresentation3
    );
    expect(
      serialize(
        {
          mapping: result.mapping,
          entities: result.entities,
          requests: {},
        },
        result.resource,
        serializeConfigPresentation3
      )
    ).toMatchSnapshot();

    expect(reserialized).toEqual(original);
  });

  describe('Collection and manifest labels', () => {
    const manifestFromCollection = {
      id: 'https://iiif.wellcomecollection.org/presentation/b18031511_0001',
      type: 'Manifest',
      label: {
        en: ['Volume 1', 'The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.'],
      },
    };
    const manifestOnItsOwn = {
      id: 'https://iiif.wellcomecollection.org/presentation/b18031511_0001',
      type: 'Manifest',
      label: {
        en: ['The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.'],
      },
    };

    test('Collection, then manifest.', () => {
      const initial = mergeEntities(
        {
          id: 'https://iiif.wellcomecollection.org/presentation/b18031511_0001',
          type: 'Manifest',
        } as any,
        manifestFromCollection,
        { parent: { id: 'https://example.org/collection-1' } }
      );
      expect(initial).toMatchSnapshot();

      const merged = mergeEntities(initial as any, manifestOnItsOwn, { parent: undefined, isTopLevel: true });
      expect(merged).toMatchSnapshot();
    });

    test('Manifest then Collection', () => {
      // Then test Manifest, then collection
      const initial = mergeEntities(
        {
          id: 'https://iiif.wellcomecollection.org/presentation/b18031511_0001',
          type: 'Manifest',
        } as any,
        manifestOnItsOwn,
        { parent: undefined, isTopLevel: true }
      );
      expect(initial).toMatchSnapshot();

      const merged = mergeEntities(initial as any, manifestFromCollection, {
        parent: { id: 'https://example.org/collection-1' },
      });
      expect(merged).toMatchSnapshot();
    });
  });
});
