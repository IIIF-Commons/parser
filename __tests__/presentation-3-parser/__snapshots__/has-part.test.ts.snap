// Vitest Snapshot v1

exports[`Has part issues > Collection and manifest labels > Collection, then manifest. 1`] = `
{
  "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
  "iiif-parser:hasPart": [
    {
      "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "iiif-parser:partOf": "https://example.org/collection-1",
      "type": "Manifest",
    },
  ],
  "label": {
    "en": [
      "Volume 1",
      "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.",
    ],
  },
  "type": "Manifest",
}
`;

exports[`Has part issues > Collection and manifest labels > Collection, then manifest. 2`] = `
{
  "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
  "iiif-parser:hasPart": [
    {
      "@explicit": true,
      "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "iiif-parser:partOf": "https://example.org/collection-1",
      "label": {
        "en": [
          "Volume 1",
          "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.",
        ],
      },
      "type": "Manifest",
    },
    {
      "@explicit": true,
      "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "iiif-parser:partOf": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "label": {
        "en": [
          "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.",
        ],
      },
      "type": "Manifest",
    },
  ],
  "label": {
    "en": [
      "Volume 1",
      "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.",
    ],
  },
  "type": "Manifest",
}
`;

exports[`Has part issues > Collection and manifest labels > Manifest then Collection 1`] = `
{
  "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
  "iiif-parser:hasPart": [
    {
      "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "iiif-parser:partOf": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "type": "Manifest",
    },
  ],
  "label": {
    "en": [
      "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.",
    ],
  },
  "type": "Manifest",
}
`;

exports[`Has part issues > Collection and manifest labels > Manifest then Collection 2`] = `
{
  "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
  "iiif-parser:hasPart": [
    {
      "@explicit": true,
      "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "iiif-parser:partOf": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "label": {
        "en": [
          "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.",
        ],
      },
      "type": "Manifest",
    },
    {
      "@explicit": true,
      "id": "https://iiif.wellcomecollection.org/presentation/b18031511_0001",
      "iiif-parser:partOf": "https://example.org/collection-1",
      "label": {
        "en": [
          "Volume 1",
          "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.",
        ],
      },
      "type": "Manifest",
    },
  ],
  "label": {
    "en": [
      "The biological basis of medicine / edited by E. Edward Bittar ; assisted by Neville Bittar.",
      "Volume 1",
    ],
  },
  "type": "Manifest",
}
`;

exports[`Has part issues > Example manifest with thumbnail ID the same as the main resource 1`] = `
{
  "entities": {
    "Agent": {},
    "Annotation": {
      "https://iiif.io/api/cookbook/recipe/0005-image-service/annotation/p0001-image": {
        "body": [
          {
            "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
            "type": "ContentResource",
          },
        ],
        "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/annotation/p0001-image",
        "iiif-parser:hasPart": [
          {
            "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/annotation/p0001-image",
            "iiif-parser:partOf": "https://iiif.io/api/cookbook/recipe/0005-image-service/page/p1/1",
            "type": "Annotation",
          },
        ],
        "motivation": [
          "painting",
        ],
        "target": {
          "source": {
            "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/canvas/p1",
            "type": "Canvas",
          },
          "type": "SpecificResource",
        },
        "type": "Annotation",
      },
    },
    "AnnotationCollection": {},
    "AnnotationPage": {
      "https://iiif.io/api/cookbook/recipe/0005-image-service/page/p1/1": {
        "behavior": [],
        "homepage": [],
        "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/page/p1/1",
        "iiif-parser:hasPart": [
          {
            "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/page/p1/1",
            "iiif-parser:partOf": "https://iiif.io/api/cookbook/recipe/0005-image-service/canvas/p1",
            "type": "AnnotationPage",
          },
        ],
        "items": [
          {
            "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/annotation/p0001-image",
            "type": "Annotation",
          },
        ],
        "label": null,
        "metadata": [],
        "provider": [],
        "rendering": [],
        "requiredStatement": null,
        "rights": null,
        "seeAlso": [],
        "service": [],
        "summary": null,
        "thumbnail": [],
        "type": "AnnotationPage",
      },
    },
    "Canvas": {
      "https://iiif.io/api/cookbook/recipe/0005-image-service/canvas/p1": {
        "accompanyingCanvas": null,
        "annotations": [],
        "behavior": [],
        "duration": 0,
        "height": 3024,
        "homepage": [],
        "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/canvas/p1",
        "items": [
          {
            "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/page/p1/1",
            "type": "AnnotationPage",
          },
        ],
        "label": {
          "en": [
            "Canvas with a single IIIF image",
          ],
        },
        "metadata": [],
        "navDate": null,
        "partOf": [],
        "placeholderCanvas": null,
        "provider": [],
        "rendering": [],
        "requiredStatement": null,
        "rights": null,
        "seeAlso": [],
        "service": [],
        "summary": null,
        "thumbnail": [
          {
            "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
            "type": "ContentResource",
          },
        ],
        "type": "Canvas",
        "width": 4032,
      },
    },
    "Collection": {},
    "ContentResource": {
      "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg": {
        "format": "image/jpeg",
        "height": 3024,
        "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
        "iiif-parser:hasPart": [
          {
            "@explicit": true,
            "format": {},
            "height": {},
            "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
            "iiif-parser:partOf": "https://iiif.io/api/cookbook/recipe/0005-image-service/annotation/p0001-image",
            "service": {},
            "type": "Image",
            "width": {},
          },
          {
            "@explicit": true,
            "format": {},
            "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
            "iiif-parser:partOf": "https://iiif.io/api/cookbook/recipe/0005-image-service/canvas/p1",
            "type": "Image",
          },
        ],
        "service": [
          {
            "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen",
            "profile": "level1",
            "type": "ImageService3",
          },
        ],
        "type": "Image",
        "width": 4032,
      },
    },
    "Manifest": {
      "https://iiif.io/api/cookbook/recipe/0005-image-service/manifest.json": {
        "@context": "http://iiif.io/api/presentation/3/context.json",
        "accompanyingCanvas": null,
        "annotations": [],
        "behavior": [],
        "homepage": [],
        "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/manifest.json",
        "iiif-parser:hasPart": [
          {
            "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/manifest.json",
            "iiif-parser:partOf": "https://iiif.io/api/cookbook/recipe/0005-image-service/manifest.json",
            "type": "Manifest",
          },
        ],
        "items": [
          {
            "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/canvas/p1",
            "type": "Canvas",
          },
        ],
        "label": {
          "en": [
            "Picture of GÃ¶ttingen taken during the 2019 IIIF Conference",
          ],
        },
        "metadata": [],
        "navDate": null,
        "partOf": [],
        "placeholderCanvas": null,
        "provider": [],
        "rendering": [],
        "requiredStatement": null,
        "rights": null,
        "seeAlso": [],
        "service": [],
        "services": [],
        "start": null,
        "structures": [],
        "summary": null,
        "thumbnail": [],
        "type": "Manifest",
        "viewingDirection": "left-to-right",
      },
    },
    "Range": {},
    "Selector": {},
    "Service": {
      "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen": {
        "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen",
        "profile": "level1",
        "type": "ImageService3",
      },
    },
  },
  "mapping": {
    "https://iiif.io/api/cookbook/recipe/0005-image-service/annotation/p0001-image": "Annotation",
    "https://iiif.io/api/cookbook/recipe/0005-image-service/canvas/p1": "Canvas",
    "https://iiif.io/api/cookbook/recipe/0005-image-service/manifest.json": "Manifest",
    "https://iiif.io/api/cookbook/recipe/0005-image-service/page/p1/1": "AnnotationPage",
    "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg": "ContentResource",
  },
  "resource": {
    "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/manifest.json",
    "type": "Manifest",
  },
}
`;

exports[`Has part issues > Example manifest with thumbnail ID the same as the main resource 2`] = `
{
  "@context": "http://iiif.io/api/presentation/3/context.json",
  "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/manifest.json",
  "items": [
    {
      "height": 3024,
      "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/canvas/p1",
      "items": [
        {
          "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/page/p1/1",
          "items": [
            {
              "body": {
                "format": "image/jpeg",
                "height": 3024,
                "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
                "service": [
                  {
                    "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen",
                    "profile": "level1",
                    "type": "ImageService3",
                  },
                ],
                "type": "Image",
                "width": 4032,
              },
              "id": "https://iiif.io/api/cookbook/recipe/0005-image-service/annotation/p0001-image",
              "motivation": "painting",
              "target": "https://iiif.io/api/cookbook/recipe/0005-image-service/canvas/p1",
              "type": "Annotation",
            },
          ],
          "type": "AnnotationPage",
        },
      ],
      "label": {
        "en": [
          "Canvas with a single IIIF image",
        ],
      },
      "thumbnail": [
        {
          "format": "image/jpeg",
          "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
          "type": "Image",
        },
      ],
      "type": "Canvas",
      "width": 4032,
    },
  ],
  "label": {
    "en": [
      "Picture of GÃ¶ttingen taken during the 2019 IIIF Conference",
    ],
  },
  "type": "Manifest",
}
`;

exports[`Has part issues > Merging the same 2 entities 1`] = `
{
  "format": "image/jpeg",
  "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
  "iiif-parser:hasPart": [
    {
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
      "iiif-parser:partOf": "https://example.org/canvas-2",
      "type": "Image",
    },
  ],
  "type": "Image",
}
`;

exports[`Has part issues > Merging the same 2 entities 2`] = `
{
  "format": "image/jpeg",
  "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
  "iiif-parser:hasPart": [
    {
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
      "iiif-parser:partOf": "https://example.org/canvas-2",
      "type": "Image",
    },
    {
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
      "iiif-parser:partOf": "https://example.org/canvas-1",
      "type": "Image",
    },
  ],
  "type": "Image",
}
`;

exports[`Has part issues > Merging the same 2 entities and then different one 1`] = `
{
  "format": "image/jpeg",
  "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
  "iiif-parser:hasPart": [
    {
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
      "iiif-parser:partOf": "https://example.org/canvas-2",
      "type": "Image",
    },
  ],
  "type": "Image",
}
`;

exports[`Has part issues > Merging the same 2 entities and then different one 2`] = `
{
  "format": "image/jpeg",
  "height": 3024,
  "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
  "iiif-parser:hasPart": [
    {
      "@explicit": true,
      "format": {},
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
      "iiif-parser:partOf": "https://example.org/canvas-2",
      "type": "Image",
    },
    {
      "@explicit": true,
      "format": {},
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
      "iiif-parser:partOf": "https://example.org/canvas-1",
      "type": "Image",
    },
    {
      "@explicit": true,
      "format": {},
      "height": {},
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
      "iiif-parser:partOf": "https://example.org/canvas-3",
      "service": {},
      "type": "Image",
      "width": {},
    },
  ],
  "service": [
    {
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen",
      "profile": "level1",
      "type": "ImageService3",
    },
  ],
  "type": "Image",
  "width": 4032,
}
`;

exports[`Has part issues > merging 2 entities 1`] = `
{
  "format": "image/jpeg",
  "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
  "iiif-parser:hasPart": [
    {
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
      "iiif-parser:partOf": "https://example.org/canvas-2",
      "type": "Image",
    },
  ],
  "type": "Image",
}
`;

exports[`Has part issues > merging 2 entities 2`] = `
{
  "format": "image/jpeg",
  "height": 3024,
  "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
  "iiif-parser:hasPart": [
    {
      "@explicit": true,
      "format": {},
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
      "iiif-parser:partOf": "https://example.org/canvas-2",
      "type": "Image",
    },
    {
      "@explicit": true,
      "format": {},
      "height": {},
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/max/0/default.jpg",
      "iiif-parser:partOf": "https://example.org/canvas-1",
      "service": {},
      "type": "Image",
      "width": {},
    },
  ],
  "service": [
    {
      "id": "https://iiif.io/api/image/3.0/example/reference/918ecd18c2592080851777620de9bcb5-gottingen",
      "profile": "level1",
      "type": "ImageService3",
    },
  ],
  "type": "Image",
  "width": 4032,
}
`;
